streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala

```diff
@@ -68,36 +69,60 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
   }
 
   def stop(processAllReceivedData: Boolean): Unit = synchronized {
-    if (eventActor == null) return // scheduler has already been stopped
-    logDebug("Stopping JobScheduler")
+    val shutdownExecutor = Executors.newFixedThreadPool(1)
+    implicit val context = ExecutionContext.fromExecutorService(shutdownExecutor)
 
-    // First, stop receiving
-    receiverTracker.stop()
+    val shutdown = Future {
+      if (eventActor == null) return // scheduler has already been stopped
+      logDebug("Stopping JobScheduler")
 
-    // Second, stop generating jobs. If it has to process all received data,
-    // then this will wait for all the processing through JobScheduler to be over.
-    jobGenerator.stop(processAllReceivedData)
+      // First, stop receiving
+      receiverTracker.stop(processAllReceivedData)
 
-    // Stop the executor for receiving new jobs
-    logDebug("Stopping job executor")
-    jobExecutor.shutdown()
+      // Second, stop generating jobs. If it has to process all received data,
+      // then this will wait for all the processing through JobScheduler to be over.
+      jobGenerator.stop(processAllReceivedData)
 
-    // Wait for the queued jobs to complete if indicated
+      // Stop the executor for receiving new jobs
+      logDebug("Stopping job executor")
+      jobExecutor.shutdown()
+
+      // Wait for the queued jobs to complete if indicated
+      val terminated = if (processAllReceivedData) {
+        jobExecutor.awaitTermination(1, TimeUnit.HOURS) // just a very large period of time
+      } else {
+        jobExecutor.awaitTermination(2, TimeUnit.SECONDS)
+      }
+      if (!terminated) {
+        jobExecutor.shutdownNow()
+      }
+      logDebug("Stopped job executor")
+
+      // Stop everything else
+      listenerBus.stop()
+      ssc.env.actorSystem.stop(eventActor)
+      eventActor = null
+    }
+
```

Comment:
I dont understand the logic here. Since we created a future, shouldnt we be waiting for the future to complete. Its non-intuitive to wait on the executor. In fact, when gracefully shutting, the necessary and sufficient condition to wait for is the completion of the future (which contains waiting for job generator and job executor to shutdown).